pkg/provisioner/ironic/dependencies.go

```diff
@@ -0,0 +1,109 @@
+package ironic
+
+import (
+	"strings"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/drivers"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+)
+
+const (
+	checkRequeueDelay = time.Second * 10
+)
+
+type ironicDependenciesChecker struct {
+	client    *gophercloud.ServiceClient
+	inspector *gophercloud.ServiceClient
+}
+
+func newIronicDependenciesChecker(client *gophercloud.ServiceClient, inspector *gophercloud.ServiceClient) *ironicDependenciesChecker {
+	return &ironicDependenciesChecker{
+		client:    client,
+		inspector: inspector,
+	}
+}
+
+func (i *ironicDependenciesChecker) IsReady() (result provisioner.Result, err error) {
+
+	ready, err := i.checkIronic()
+	if ready && err == nil {
+		ready, err = i.checkIronicInspector()
+	}
+
+	if !ready && err == nil {
+		result.Dirty = true
+		result.RequeueAfter = checkRequeueDelay
+	}
+
+	return result, err
+}
+
+func (i *ironicDependenciesChecker) checkEndpoint(client *gophercloud.ServiceClient) (ready bool, err error) {
+
+	// NOTE: Some versions of Ironic inspector returns 404 for /v1/ but 200 for /v1,
+	// which seems to be the default behavior for Flask. Remove the trailing slash
+	// from the client endpoint.
+	endpoint := strings.TrimSuffix(client.Endpoint, "/")
+
+	_, err = client.Get(endpoint, nil, nil)
+	ready = err == nil
+	if !ready {
+		switch err.(type) {
+		case
+			gophercloud.ErrUnexpectedResponseCode,
```

Comment:
What types of errors might be returned that wouldn't be filtered out by this switch?

I think what this is trying to do is turn an error caused by the fact the pod is still starting up into a less obtrusive "not ready" condition. That makes sense. I'm not sure why we wouldn't treat all errors in the same way, always. It seems like if the service is returning an error when we make simple status check calls like this, then it is not ready. We could just log the error at info level and report not ready and try again in a while.

Reply:
All the other usual net errors (amongst the most common ones, I'd say for example `net/url.Error` when the service is not present at all, ie during the bootstrap), some additional wrapped http related errors and JSON related errors (mostly related to decoding). I agree that all this kind of errors are indicating anyhow that the service is not ready/available, so we could simplify even more the code.